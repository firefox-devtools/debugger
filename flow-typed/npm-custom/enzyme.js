/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// Using this custom definition of the enzyme module avoids generating some
// flow errors that are unrelated to uses of enzyme itself...
// It would be nice to clean this up.
declare module "enzyme" {
  declare module.exports: any;
}
