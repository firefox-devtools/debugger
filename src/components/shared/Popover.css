.popover {
  position: fixed;
  z-index: 100;
}

.popover .gap {
  height: 5px;
  padding-top: 5px;
}

.popover::before,
.popover::after {
  content: '';
  height: 0;
  width: 0;
  position: absolute;
  border: 7px solid transparent;
  left: calc(20% - 7px); /* corresponds to calculation in Popover.js */
}

.popover::before {
  border-bottom-color: var(--theme-splitter-color);
  top: -9px;
}

.theme-dark .popover::before {
  border-bottom-color: var(--theme-body-color);
}

.popover::after {
  border-bottom-color: var(--theme-body-background);
  top: -8px;
}

.popover.up::before {
  border-bottom-color: transparent;
  border-top-color: var(--theme-splitter-color);
  bottom: -9px;
  top: initial;
}

.theme-dark .popover.up::before {
  border-top-color: var(--theme-body-color);
}

.popover.up::after {
  border-bottom-color: transparent;
  border-top-color: var(--theme-body-background);
  bottom: -8px;
  top: initial;
}
