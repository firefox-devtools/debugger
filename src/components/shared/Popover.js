/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow
import React, { Component } from "react";
import classNames from "classnames";
import BracketArrow from "./BracketArrow";

import "./Popover.css";

type Props = {
  editorRef: ?HTMLDivElement,
  targetPosition: Object,
  children: ?React$Element<any>,
  onPopoverCoords: Function,
  onMouseLeave: (e: SyntheticMouseEvent<HTMLDivElement>) => void,
  onKeyDown: (e: KeyboardEvent) => void,
  type?: "popover" | "tooltip"
};

type Orientation = "up" | "down" | "right";
type TargetMid = {
  x: number,
  y: number
};
export type Coords = {
  left: number,
  top: number,
  targetMid: TargetMid,
  orientation: Orientation
};

type State = { coords: Coords };

class Popover extends Component<Props, State> {
  $popover: ?HTMLDivElement;
  $tooltip: ?HTMLDivElement;
  state = {
    coords: {
      left: 0,
      top: 0,
      orientation: "down",
      targetMid: { x: 0, y: 0 }
    }
  };

  static defaultProps = {
    onMouseLeave: () => {},
    onPopoverCoords: () => {},
    onKeyDown: () => {},
    type: "popover"
  };

  componentDidMount() {
    const { type } = this.props;
    const coords =
      type == "popover" ? this.getPopoverCoords() : this.getTooltipCoords();

    if (coords) {
      this.setState({ coords });
    }

    this.props.onPopoverCoords(coords);
  }

  calculateLeft(
    target: ClientRect,
    editor: ClientRect,
    popover: ClientRect,
    orientation?: Orientation
  ) {
    const estimatedLeft = target.left;
    const estimatedRight = estimatedLeft + popover.width;
    const isOverflowingRight = estimatedRight > editor.right;
    if (orientation === "right") {
      return target.left + target.width + 5;
    }
    if (isOverflowingRight) {
      const adjustedLeft = editor.right - popover.width - 8;
      return adjustedLeft;
    }
    return estimatedLeft;
  }

  calculateTopForRightOrientation = (
    target: ClientRect,
    editor: ClientRect,
    popover: ClientRect
  ) => {
    if (popover.height < editor.height) {
      const rightOrientationTop = target.top - popover.height / 2;
      if (rightOrientationTop < editor.top) {
        return editor.top;
      }
      const rightOrientationBottom = rightOrientationTop + popover.height;
      if (rightOrientationBottom > editor.bottom) {
        return editor.bottom - popover.height;
      }
      return rightOrientationTop;
    }
    return 0;
  };

  calculateOrientation(
    target: ClientRect,
    editor: ClientRect,
    popover: ClientRect
  ) {
    const estimatedBottom = target.bottom + popover.height;
    if (editor.bottom > estimatedBottom) {
      return "down";
    }
    const upOrientationTop = target.top - popover.height;
    if (upOrientationTop > editor.top) {
      return "up";
    }

    return "right";
  }

  calculateTop = (
    target: ClientRect,
    editor: ClientRect,
    popover: ClientRect,
    orientation?: string
  ) => {
    if (orientation === "down") {
      return target.bottom;
    }
    if (orientation === "up") {
      return target.top - popover.height;
    }

    return this.calculateTopForRightOrientation(target, editor, popover);
  };

  getPopoverCoords() {
    if (!this.$popover || !this.props.editorRef) {
      return null;
    }

    const popover = this.$popover;
    const editor = this.props.editorRef;
    const popoverRect = popover.getBoundingClientRect();
    const editorRect = editor.getBoundingClientRect();
    const targetRect = this.props.targetPosition;
    const orientation = this.calculateOrientation(
      targetRect,
      editorRect,
      popoverRect
    );
    const top = this.calculateTop(
      targetRect,
      editorRect,
      popoverRect,
      orientation
    );
    const popoverLeft = this.calculateLeft(
      targetRect,
      editorRect,
      popoverRect,
      orientation
    );
    let targetMid;
    if (orientation === "right") {
      targetMid = {
        x: -14,
        y: targetRect.top - top - 2
      };
    } else {
      targetMid = {
        x: targetRect.left - popoverLeft + targetRect.width / 2 - 8,
        y: 0
      };
    }

    return {
      left: popoverLeft,
      top,
      orientation,
      targetMid
    };
  }

  getTooltipCoords() {
    if (!this.$tooltip || !this.props.editorRef) {
      return null;
    }
    const tooltip = this.$tooltip;
    const editor = this.props.editorRef;
    const tooltipRect = tooltip.getBoundingClientRect();
    const editorRect = editor.getBoundingClientRect();
    const targetRect = this.props.targetPosition;
    const left = this.calculateLeft(targetRect, editorRect, tooltipRect);
    const enoughRoomForTooltipAbove =
      targetRect.top - editorRect.top > tooltipRect.height;
    const top = enoughRoomForTooltipAbove
      ? targetRect.top - tooltipRect.height
      : targetRect.bottom;

    return {
      left,
      top,
      orientation: enoughRoomForTooltipAbove ? "up" : "down",
      targetMid: { x: 0, y: 0 }
    };
  }

  getChildren() {
    const { children } = this.props;
    const { orientation } = this.state.coords;
    const gap = <div className="gap" key="gap" />;
    return orientation === "up" ? [children, gap] : [gap, children];
  }

  getPopoverArrow(orientation: Orientation, left: number, top: number) {
    const arrowProps = {};
    if (orientation === "up") {
      Object.assign(arrowProps, { orientation: "down", bottom: 5, left });
    } else if (orientation === "down") {
      Object.assign(arrowProps, { orientation: "up", top: -7, left });
    } else {
      Object.assign(arrowProps, { orientation: "left", top, left: -9 });
    }

    return <BracketArrow {...arrowProps} />;
  }

  renderPopover() {
    const { top, left, orientation, targetMid } = this.state.coords;
    const { onMouseLeave, onKeyDown } = this.props;
    const arrow = this.getPopoverArrow(orientation, targetMid.x, targetMid.y);

    return (
      <div
        className={classNames("popover", `orientation-${orientation}`, {
          up: orientation === "up"
        })}
        onMouseLeave={onMouseLeave}
        onKeyDown={onKeyDown}
        style={{ top, left }}
        ref={c => (this.$popover = c)}
      >
        {arrow}
        {this.getChildren()}
      </div>
    );
  }

  renderTooltip() {
    const { top, left } = this.state.coords;
    const { onMouseLeave, onKeyDown } = this.props;
    return (
      <div
        className="tooltip"
        onMouseLeave={onMouseLeave}
        onKeyDown={onKeyDown}
        style={{ top, left }}
        ref={c => (this.$tooltip = c)}
      >
        {this.getChildren()}
      </div>
    );
  }

  render() {
    const { type } = this.props;

    if (type === "tooltip") {
      return this.renderTooltip();
    }

    return this.renderPopover();
  }
}

export default Popover;
