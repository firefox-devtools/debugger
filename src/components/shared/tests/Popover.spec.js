/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

import React from "react";
import { mount, shallow } from "enzyme";

import Popover from "../Popover";

describe("Popover", () => {
  const onMouseLeave = jest.fn();
  const editorRef = {
    getBoundingClientRect() {
      return {
        x: 0,
        y: 0,
        width: 100,
        height: 100,
        top: 250,
        right: 0,
        bottom: 0,
        left: 20
      };
    }
  };
  const targetPosition = {
    x: 100,
    y: 200,
    width: 300,
    height: 300,
    top: 50,
    right: 0,
    bottom: 0,
    left: 200
  };
  const popover = shallow(
    <Popover
      onMouseLeave={onMouseLeave}
      editorRef={editorRef}
      targetPosition={targetPosition}
    >
      <h1>Poppy!</h1>
    </Popover>
  );
  const tooltip = shallow(
    <Popover
      type="tooltip"
      onMouseLeave={onMouseLeave}
      editorRef={editorRef}
      targetPosition={targetPosition}
    >
      <h1>Toolie!</h1>
    </Popover>
  );
  const div = document.createElement("div");
  const event = { currentTarget: div };
  beforeEach(() => onMouseLeave.mockClear());
  it("render", () => expect(popover).toMatchSnapshot());
  it("render (tooltip)", () => expect(tooltip).toMatchSnapshot());
  it("calls mouseLeave", () => {
    popover.find(".popover").simulate("mouseleave", event);
    expect(onMouseLeave).toHaveBeenCalled();
  });
  it("calls mouseLeave (tooltip)", () => {
    tooltip.find(".tooltip").simulate("mouseleave", event);
    expect(onMouseLeave).toHaveBeenCalled();
  });
  it("no mouse leave on bracket or gap", () => {
    div.className = "bracket-arrow";
    popover.find(".popover").simulate("mouseleave", event);
    expect(onMouseLeave).not.toHaveBeenCalled();
  });
  it("mount popover", () => {
    const mountedPopover = mount(
      <Popover
        onMouseLeave={onMouseLeave}
        editorRef={editorRef}
        targetPosition={targetPosition}
      >
        <h1>Poppy!</h1>
      </Popover>
    );
    expect(mountedPopover).toMatchSnapshot();
  });
  it("mount tooltip", () => {
    const mountedTooltip = mount(
      <Popover
        type="tooltip"
        onMouseLeave={onMouseLeave}
        editorRef={editorRef}
        targetPosition={targetPosition}
      >
        <h1>Toolie!</h1>
      </Popover>
    );
    expect(mountedTooltip).toMatchSnapshot();
  });
});
