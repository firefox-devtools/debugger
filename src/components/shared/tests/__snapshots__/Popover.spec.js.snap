// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover mount popover 1`] = `
<Popover
  editorRef={
    Object {
      "getBoundingClientRect": [Function],
    }
  }
  onKeyDown={[MockFunction]}
  onMouseLeave={[MockFunction]}
  onPopoverCoords={[Function]}
  targetPosition={
    Object {
      "bottom": 0,
      "height": 300,
      "left": 200,
      "right": 0,
      "top": 50,
      "width": 300,
      "x": 100,
      "y": 200,
    }
  }
  type="popover"
>
  <div
    className="popover orientation-right"
    onKeyDown={[MockFunction]}
    onMouseLeave={[MockFunction]}
    style={
      Object {
        "left": 505,
        "top": 250,
      }
    }
  >
    <BracketArrow
      left={-9}
      orientation="left"
      top={-202}
    >
      <div
        className="bracket-arrow left"
        style={
          Object {
            "bottom": undefined,
            "left": -9,
            "top": -202,
          }
        }
      />
    </BracketArrow>
    <div
      className="gap"
      key="gap"
    />
    <h1>
      Poppy!
    </h1>
  </div>
</Popover>
`;

exports[`Popover mount tooltip 1`] = `
<Popover
  editorRef={
    Object {
      "getBoundingClientRect": [Function],
    }
  }
  onKeyDown={[MockFunction]}
  onMouseLeave={[MockFunction]}
  onPopoverCoords={[Function]}
  targetPosition={
    Object {
      "bottom": 0,
      "height": 300,
      "left": 200,
      "right": 0,
      "top": 50,
      "width": 300,
      "x": 100,
      "y": 200,
    }
  }
  type="tooltip"
>
  <div
    className="tooltip"
    onKeyDown={[MockFunction]}
    onMouseLeave={[MockFunction]}
    style={
      Object {
        "left": -8,
        "top": 0,
      }
    }
  >
    <div
      className="gap"
      key="gap"
    />
    <h1>
      Toolie!
    </h1>
  </div>
</Popover>
`;

exports[`Popover render (tooltip) 1`] = `
<div
  className="tooltip"
  onKeyDown={[MockFunction]}
  onMouseLeave={[MockFunction]}
  style={
    Object {
      "left": 0,
      "top": 0,
    }
  }
>
  <div
    className="gap"
    key="gap"
  />
  <h1>
    Toolie!
  </h1>
</div>
`;

exports[`Popover render 1`] = `
<Popover
  editorRef={
    Object {
      "getBoundingClientRect": [Function],
    }
  }
  onKeyDown={[MockFunction]}
  onMouseLeave={[MockFunction]}
  onPopoverCoords={[Function]}
  targetPosition={
    Object {
      "bottom": 0,
      "height": 300,
      "left": 200,
      "right": 0,
      "top": 50,
      "width": 300,
      "x": 100,
      "y": 200,
    }
  }
  type="popover"
>
  <div
    className="popover orientation-right"
    onKeyDown={[MockFunction]}
    onMouseLeave={[MockFunction]}
    style={
      Object {
        "left": 505,
        "top": 250,
      }
    }
  >
    <BracketArrow
      left={-9}
      orientation="left"
      top={-202}
    >
      <div
        className="bracket-arrow left"
        style={
          Object {
            "bottom": undefined,
            "left": -9,
            "top": -202,
          }
        }
      />
    </BracketArrow>
    <div
      className="gap"
      key="gap"
    />
    <h1>
      Poppy!
    </h1>
  </div>
</Popover>
`;
